cmd/observe/io_reader_observer.go

```diff
@@ -0,0 +1,103 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package observe
+
+import (
+	"bufio"
+	"context"
+	"io"
+
+	"github.com/cilium/cilium/api/v1/observer"
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	"github.com/cilium/cilium/pkg/hubble/filters"
+	"github.com/cilium/hubble/pkg/logger"
+	"github.com/golang/protobuf/jsonpb"
+	"google.golang.org/grpc"
+)
+
+// ioReaderObserver implements ObserverClient interface. It reads flows
+// in jsonpb format from an io.Reader.
+type ioReaderObserver struct {
+	reader *bufio.Reader
+}
+
+func newIOReaderObserver(reader io.Reader) *ioReaderObserver {
+	return &ioReaderObserver{
+		reader: bufio.NewReader(reader),
+	}
+}
+
+func (o *ioReaderObserver) GetFlows(_ context.Context, in *observer.GetFlowsRequest, _ ...grpc.CallOption) (observer.Observer_GetFlowsClient, error) {
+	return newIOReaderClient(o.reader, in)
+}
+
+func (o *ioReaderObserver) GetNodes(_ context.Context, _ *observer.GetNodesRequest, _ ...grpc.CallOption) (*observer.GetNodesResponse, error) {
+	panic("GetNodes is not supported")
+}
+
+func (o *ioReaderObserver) ServerStatus(_ context.Context, _ *observer.ServerStatusRequest, _ ...grpc.CallOption) (*observer.ServerStatusResponse, error) {
+	panic("ServerStatus is not supported")
+}
+
+// ioReaderClient implements Observer_GetFlowsClient.
+type ioReaderClient struct {
+	reader  *bufio.Reader
+	request *observer.GetFlowsRequest
+	allow   filters.FilterFuncs
+	deny    filters.FilterFuncs
+	grpc.ClientStream
+}
+
+func newIOReaderClient(reader *bufio.Reader, request *observer.GetFlowsRequest) (*ioReaderClient, error) {
+	allow, err := filters.BuildFilterList(context.Background(), request.Whitelist, filters.DefaultFilters)
+	if err != nil {
+		return nil, err
+	}
+	deny, err := filters.BuildFilterList(context.Background(), request.Blacklist, filters.DefaultFilters)
+	if err != nil {
+		return nil, err
+	}
+	return &ioReaderClient{
+		reader:  reader,
+		request: request,
+		allow:   allow,
+		deny:    deny,
+	}, nil
+}
+
+func (c *ioReaderClient) Recv() (*observer.GetFlowsResponse, error) {
+	for {
+		line, err := c.reader.ReadString('\n')
+		if err != nil {
+			return nil, err
+		}
+		var res observer.GetFlowsResponse
+		err = jsonpb.UnmarshalString(line, &res)
+		if err != nil {
+			logger.Logger.WithError(err).WithField("line", line).Warn("Failed to unmarshal json to flow")
+			continue
+		}
+		if c.request.Since != nil && c.request.Since.AsTime().After(res.Time.AsTime()) {
+			continue
+		}
+		if c.request.Until != nil && c.request.Until.AsTime().Before(res.Time.AsTime()) {
+			continue
+		}
```

Comment:
This doesn't handle `c.request.Number`. Is that intentional?

Reply:
yes it is somewhat intentional. you can always do this instead 😉 
```
hubble observe -n kube-system | tail -n 20
```

Reply:
How would this work for our Windows users though? 😀

Reply:
that's a good question 😱 not tested, but with powershell something like this might work?
```
gc flow.log | hubble observe -n kube-system | select -last 20
```